### PR TITLE
Feat(pipelineloop): Break task as custom task.

### LINKED ISSUE #833: Convert pipeline loop Break task into custom task.
/kind feature

**Description:**
Currently we use the break task as a placeholder to tell the pipeline loop controller to break the loop and cancel the running pipelines. We want to see can we make the break task into a dummy custom task, so we don't have to waste resources on creating a dummy pod.

**Additional information:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by ScrapCodes**
**Which issue is resolved by this Pull Request:** 
Resolves #833 

**Description of your changes:**
Now break task is also owned and controlled by pipelineloop controller. A Break task is a dummy task used as a marker for breaking the loop. A break task custom task saves resources as new real container or pod is started to process it.

**Environment tested:**

* Python Version (use `python --version`):
* Tekton Version (use `tkn version`):
* Kubernetes Version (use `kubectl version`):
* OS (e.g. from `/etc/os-release`):

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go [hunk 1]
```diff
@@ -689,6 +689,17 @@ func (c *Reconciler) updatePipelineRunStatus(ctx context.Context, run *v1alpha1.
 					return 0, nil, nil, fmt.Errorf("could not cancel PipelineRuns belonging to Run %s."+
 						" %#v", run.Name, err)
 				}
+				// Cancel the break task.
+				runBreakTask, err := c.pipelineClientSet.TektonV1alpha1().Runs(run.Namespace).Get(ctx, runStatus.PipelineTaskName, metav1.GetOptions{})
+				if err != nil {
+					logger.Errorf("could not cancel the break task run, %v", err)
+				}
+				runBreakTask.Status.InitializeConditions()
+				runBreakTask.Status.MarkRunFailed(v1beta1.PipelineRunReasonCancelled.String(), "Break task is a dummy task")
```
Comment (Tomcli) [inline tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go:698]: Maybe try to mark it as completed?

**Comment by Tomcli** [comment]
I try to run this example from @typeness and converted the break task to custom task. However, I see the custom task is stuck with an empty status. @ScrapCodes Can you help debug this?

Here is the python code:
```
from kfp import dsl, components
from kfp.components import load_component_from_text
from kfp_tekton import tekton
MY_CUSTOM_TASK_IMAGE_NAME = "veryunique/image:latest"
from kfp_tekton.tekton import TEKTON_CUSTOM_TASK_IMAGES, CEL_ConditionOp
TEKTON_CUSTOM_TASK_IMAGES = TEKTON_CUSTOM_TASK_IMAGES.append(MY_CUSTOM_TASK_IMAGE_NAME)


PrintOp = load_component_from_text("""
  name: print
  inputs:
  - name: msg
  outputs:
  - name: stdout
  implementation:
    container:
      image: alpine:3.6
      command:
      - concat:
        - "echo "
        - { inputValue: msg }
""")

op1_yaml = '''\
name: 'my-in-coop1'
inputs:
- {name: item, type: Integer}
- {name: param, type: Integer}
implementation:
    container:
        image: library/bash:4.4.23
        command: ['sh', '-c']
        args:
        - |
          set -e
          echo op1 "$0" "$1"
        - {inputValue: item}
        - {inputValue: param}
'''

break_yaml = '''\
name: 'pipelineloop-break-operation'
description: 'Break Operation using PipelineLoop'
implementation:
    container:
        image: %s
        command:
        - any
        - command
        args:
        - --apiVersion
        - custom.tekton.dev/v1alpha1
        - --kind
        - BreakTask
        - --name
        - pipelineloop-break-operation
        - --taskSpec
        - |
          {"placeholder": "placeholder"}
''' % MY_CUSTOM_TASK_IMAGE_NAME


@dsl.pipeline(name='pipeline')
def pipeline(param: int = 10):
    loop_args = [1, 2]
    with dsl.ParallelFor(loop_args, parallelism=1) as item:
        op1_template = components.load_component_from_text(op1_yaml)
        op1 = op1_template(item, param)
        condi_1 = tekton.CEL_ConditionOp(f"{item} == 2").output
        with dsl.Condition(condi_1 == 'true'):
            break_template = components.load_component_from_text(break_yaml)
            break_template()

if __name__ == '__main__':
  from kfp_tekton.compiler import TektonCompiler as Compiler
  from kfp_tekton.compiler.pipeline_utils import TektonPipelineConf
  tekton_pipeline_conf = TektonPipelineConf()
  tekton_pipeline_conf.set_tekton_inline_spec(True)
  # tekton_pipeline_conf.set_resource_in_separate_yaml(True)
  Compiler().compile(pipeline, __file__.replace('.py', '.yaml'), tekton_pipeline_conf=tekton_pipeline_conf)
```

Here is the yaml:
```
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: pipeline
  annotations:
    tekton.dev/output_artifacts: '{}'
    tekton.dev/input_artifacts: '{}'
    tekton.dev/artifact_bucket: mlpipeline
    tekton.dev/artifact_endpoint: minio-service.kubeflow:9000
    tekton.dev/artifact_endpoint_scheme: http://
    tekton.dev/artifact_items: '{"my-in-coop1": []}'
    sidecar.istio.io/inject: "false"
    pipelines.kubeflow.org/big_data_passing_format: $(workspaces.$TASK_NAME.path)/artifacts/$ORIG_PR_NAME/$TASKRUN_NAME/$TASK_PARAM_NAME
    pipelines.kubeflow.org/pipeline_spec: '{"inputs": [{"default": "10", "name": "param",
      "optional": true, "type": "Integer"}], "name": "pipeline"}'
spec:
  params:
  - name: param
    value: '10'
  pipelineSpec:
    params:
    - name: param
      default: '10'
    tasks:
    - name: pipeline-490d3-for-loop-2
      params:
      - name: loop-item-param-1
        value: '[1, 2]'
      - name: param
        value: $(params.param)
      taskSpec:
        apiVersion: custom.tekton.dev/v1alpha1
        kind: PipelineLoop
        spec:
          pipelineSpec:
            params:
            - name: loop-item-param-1
              type: string
            - name: param
              type: string
            tasks:
            - name: my-in-coop1
              params:
              - name: loop-item-param-1
                value: $(params.loop-item-param-1)
              - name: param
                value: $(params.param)
              taskSpec:
                steps:
                - name: main
                  args:
                  - |
                    set -e
                    echo op1 "$0" "$1"
                  - $(inputs.params.loop-item-param-1)
                  - $(inputs.params.param)
                  command:
                  - sh
                  - -c
                  image: library/bash:4.4.23
                params:
                - name: loop-item-param-1
                  type: string
                - name: param
                  type: string
                metadata:
                  labels:
                    pipelines.kubeflow.org/pipelinename: ''
                    pipelines.kubeflow.org/generation: ''
                    pipelines.kubeflow.org/cache_enabled: "true"
                  annotations:
                    pipelines.kubeflow.org/component_spec_digest: '{"name": "my-in-coop1",
                      "outputs": [], "version": "my-in-coop1@sha256=8ccab3a28a39a406554d964865f2ccb0aed854a43b6de827f613eff2bccd6f8f"}'
                    tekton.dev/template: ''
              timeout: 525600m
            - name: condition-cel
              params:
              - name: outcome
                value: $(params.loop-item-param-1) == 2
              taskRef:
                name: cel_condition
                apiVersion: cel.tekton.dev/v1alpha1
                kind: CEL
              timeout: 525600m
            - name: pipelineloop-break-operation
              taskSpec:
                apiVersion: custom.tekton.dev/v1alpha1
                kind: BreakTask
                spec:
                  placeholder: placeholder
                metadata:
                  labels:
                    pipelines.kubeflow.org/pipelinename: ''
                    pipelines.kubeflow.org/generation: ''
                    pipelines.kubeflow.org/cache_enabled: "true"
                  annotations:
                    pipelines.kubeflow.org/component_spec_digest: '{"name": "pipelineloop-break-operation",
                      "outputs": [], "version": "pipelineloop-break-operation@sha256=b03bca96792e1885039ad6ac5cc13de537b3862a5e09697f2552c71968b3aac8"}'
                    tekton.dev/template: ''
              when:
              - input: $(tasks.condition-cel.results.outcome)
                operator: in
                values:
                - "true"
              timeout: 525600m
          parallelism: 1
          iterateParam: loop-item-param-1
  timeout: 525600m
```

**Comment by Tomcli** [comment]
I got this 
```
$ kubectl get run -n kubeflow
pi7e9a7f3707ffa375b16f11c65f0ed222-pipelineloop-break-operation
pipeline-27f27-pi4c0551b227d07383bc0b16fbcf683bd1-condition-cel   True        EvaluationSuccess   12m         12m
pipeline-27f27-pi7e9a7f3707ffa375b16f11c65f0ed222-condition-cel   True        EvaluationSuccess   12m         12m
pipeline-27f27-pipeline-490d3-for-loop-2                          True        Succeeded           12m         12m
```

**Comment by Tomcli** [comment]
Also, can the `BreakTask` be referred with `taskRef` without creating a `BreakTask` Custom resource? Similar to how CEL doesn't need a CR to use `taskRef`

**Comment by ScrapCodes** [comment]
Hi @Tomcli , somehow this approach is mysteriously not working. Even after successfully updating the break-task run resource i.e. no error, it does not show the status as updated.

**Comment by Tomcli** [comment]
Thanks @ScrapCodes, functional wise it works. However, I do notice there's a significant delay for the breaktask `run` to update its status. Is it expected?
```
pieb12882cf27cb0a7c94593351e66bdb1-pipelineloop-break-operation   True        Succeeded           11s         11s
pipeline-61bb2-pi9fd5659b4e310c56f389e726df5fa7b0-condition-cel   True        EvaluationSuccess   51s         51s
pipeline-61bb2-pieb12882cf27cb0a7c94593351e66bdb1-condition-cel   True        EvaluationSuccess   46s         45s
pipeline-61bb2-pipeline-916d8-for-loop-2                          True        Succeeded           51s         45s
```

Here the break task updated the status 34s after the pipeline is completed.

**Comment by ScrapCodes** [comment]
It is not expected behaviour, trying out with a k8s cluster on IBM Cloud, I got :
```
 k get runs -w
NAME                                                           SUCCEEDED   REASON              STARTTIME   COMPLETIONTIME
pipeline-pipeline-490d3-for-loop-2                             Unknown     Running             5s          
pipeline-pipeline-490d3-for-loop-2-00001-hg8h2-condition-cel   True        EvaluationSuccess   5s          5s
pipeline-pipeline-490d3-for-loop-2-00002-4mwhd-condition-cel                                               
pipeline-pipeline-490d3-for-loop-2                             Unknown     Running             6s          
pipeline-pipeline-490d3-for-loop-2-00002-4mwhd-condition-cel   True        EvaluationSuccess   0s          0s
pi0a6f4eecb9eff0629768a9efeb4b9bb9-pipelineloop-break-operation                                               
pipeline-pipeline-490d3-for-loop-2                                Unknown     Running             7s          
pi0a6f4eecb9eff0629768a9efeb4b9bb9-pipelineloop-break-operation   True        Succeeded           0s          0s
pipeline-pipeline-490d3-for-loop-2                                Unknown     Running             7s          
pi0a6f4eecb9eff0629768a9efeb4b9bb9-pipelineloop-break-operation   True        Succeeded           0s          0s
pipeline-pipeline-490d3-for-loop-2                                True        Succeeded           7s          0s
pi0a6f4eecb9eff0629768a9efeb4b9bb9-pipelineloop-break-operation   True        Succeeded           0s          0s
pipeline-pipeline-490d3-for-loop-2-00002-4mwhd-condition-cel      True        EvaluationSuccess   1s          1s
pipeline-pipeline-490d3-for-loop-2-00002-4mwhd-condition-cel      True        EvaluationSuccess   2s          1s

```
Unable to notice the same behaviour .

**Comment by Tomcli** [comment]
I notice it could happen during heavy workloads. It makes sense to not hold back to pipelinerun complete status since the break task event could queue up in the very end. Just something we have to note in our README so other users are aware of it.

**Comment by Tomcli** [comment]
Let's describe the break task event behavior in our SDK PR for issue #856, so users can be aware the break task status might be updated later on.

/lgtm
/approve

**Pull request merged**